### ISSUE TITLE
The ultimate refactor

### ISSUE THREAD

**Issue by adamgayoso** [YosefLab/scVI#654]
Things are getting a bit disjointed around the codebase with all the model additions. I think we can pretty straightforwardly make some abstractions that will go along way in making it easier to make models and for people to access posterior quantities.

Right now, we define a scVI deep generative model with a `Model`, a `Trainer`, and a `Posterior`. I think as of now we have done a good job abstracting the Trainer, as seen in totalVI and gimVI. The model and the posterior are a bit of a different story. Here is what I propose.

First, we refactor `Posterior` to be a `scviDataLoader` (name TBD, `scDL` for short). The scDL is essentially a wrapper on the dataloader of pytorch. It provides access to the following properties: `elbo`, `reconstruction_loss`, `indices`, etc, as well as functions that help compute these quantities, which are ultimately model specific. Therefore, there is a base scDL that is inherited for each model (scVI, totalVI, etc.). With this refactor, we remove all posterior `get` functions, DE functions, etc and will place them in a to be described class.
Recall that each scDL subclass is associated with it's own collate function (loads possibly model specific tensors from the dataset, like in totalVI)

Second, we refactor each model such that there is some base model class with the following functions...
1. `forward` like now
2. `inference` gets split into `likelihood` and `posterior`, each with the same signature as forward for downstream ease of use. `forward` calls `likelihood` and `posterior`. Each of these functions returns a dictionary with keys naming each quantity. 
3. All other model methods should not be necessary.

Third, we create a new `Posterior` class.
1. The new `Posterior` object takes a `scDL` and a model, or just the trainer, which has access to both of these.  The new `Posterior` iterates over the `scDL` and retrieves desired quantities, which come from either `likelihood` or `posterior` (which is run with inputs from `posterior`). As model functions have the expected signature, there's no need to worry about future models accepting all sorts of different tensors. Somewhere we register the names of the available quantities that can be retrieved and a helper function prints what's available.
2. The new `Posterior` can be subclassed to add special functions. For example, in totalVI we use a function of posterior predictive likelihood parameters as background free protein expression. Such functions can be added to the subclassed `Posterior`. Alternatively, they can also go in the model and would somehow get registered (haven't thought this all out yet).
3. DE is a new class. It takes a new `Posterior`, retrieves samples, and does it's magic (mostly shuffling the code we already have). It's general enough now to work with any other model parameters (e.g., differential ZI).

I'm opening this issue so we can discuss and hopefully converge on something.

**Comment by fritzo** [comment]
Hi @adamgayoso, yesterday @romain-lopez and I discussed a possible refactoring to use [Pyro](https://github.com/pyro-ppl/pyro) in scVI. Since it seems related I'll comment here, but let me know if you'd prefer to move this discussion to a separate issue.

One pattern that has evolved to work pretty well in Pyro is the following:
1. The encoder and decoder are each separate `nn.Module` objects. We call these the `model` and `guide`, respectively. Their `.forward()` methods have the same signature. In each application, users write a custom `model,guide` pair.
2. The guide encapsulates posterior terms, and the model encapsulates prior and likelihood terms of the ELBO. These terms are not exposed directly, rather we use effect handlers to [trace](http://docs.pyro.ai/en/stable/poutine.html#pyro.poutine.handlers.trace) execution of the model and guide, and can then extract samples from various random variables, i.e. "support a wide variety of downstream queries". In fact the `model` and `guide` typically return `None`; all access to random variables is by tracing `pyro.sample` statements in the model and guide.
3. For a particular class of applications, we bundle a `model,guide` pair into another `nn.Module` object. We have a generic [Predictive](http://docs.pyro.ai/en/stable/inference_algos.html#pyro.infer.predictive.Predictive) class similar to your suggested `Posterior` (indeed it was formerly named `Posterior`) that only runs the `model,guide` pair. But we also have more customized ways to bundle `model,guide` pairs, e.g. [Forecaster](http://docs.pyro.ai/en/stable/contrib.forecast.html#pyro.contrib.forecast.forecaster.Forecaster) is responsible for both training (including data subsampling) and posterior prediction.

It's unclear whether any of this might simplify scVI, but especially the tracing of `pyro.sample` sites has proved to be a flexible way to support gathering of statistics, including arbitrary conditional statistics using a [condition](http://docs.pyro.ai/en/stable/poutine.html#pyro.poutine.handlers.condition) operator or causal interventions using a [do](http://docs.pyro.ai/en/stable/poutine.html#pyro.poutine.handlers.do) operator.

cc @eb8680 @martinjankowiak

**Comment by jamestwebber** [comment]
To comment because I was also in that meeting: if you're considering a big refactor I'd definitely recommend looking at Pyro (as a biased outside observer). It will take some work to see if there are any missing features, but I _think_ that Pyro can handle all of the models and likelihoods that you're using right now in scVI, and spare you the pain of maintaining that code.

In my experience playing around with Pyro-adapted scVI code (see [czbiohub/drvish](https://github.com/czbiohub/drvish) for my code), it becomes much simpler to extend the model with additional parameters and likelihoods when you can plug in another `nn.Module` and let Pyro handle the sampling.

It's also easier to swap out likelihoods (e.g. NB vs Poisson) and it would decouple some of the DL engineering from the model (e.g. I wanted to swap out and tune the optimizer I was using, which is currently quite hard in scVI).

I proposed this a _long_ time ago (#191) but it didn't make sense then. Not that you're considering an Ultimate Refactor™ I'd look into it now. I'd be happy to contribute to that effort when I have the time!

**Comment by adamgayoso** [comment]
Thanks for your response @fritzo. I think Pyro would be a really cool option. I think the model and guide are a really nice paradigm. There also seems to be some opportunity to use Poutine in some instances. That said, I do have some concerns that may stem from my misunderstanding of what Pyro can do. I would be happy to participate in future meetings.

That said, I think what I propose in this issue (perhaps the Penultimate Refactor) would be a first step for the Ultimate refactor. Indeed, it's more of a code reshuffling, since what we call the Posterior now has too many functions. On the other hand, I'm not sure our users would like multiple API changes. Therefore, I think I can be swayed with some further understanding of Pyro.

**Comment by adamgayoso** [comment]
We have converged on something that we are working on in our various PRs. Unfortunately I don't believe we'll move over to pyro at the moment.

Our first priority was making `scvi` easier to use. In #723 we are working on just this, which essentially abstracts the user from training and interacting with pytorch. Then we will clean up all the torch modules so that it's easier to make new models and try new optimizers, etc.

**Comment by adamgayoso** [comment]
I just wanted to follow up on this, as we progress the codebase, we will mostly switch over to pytorch lightning. However, we believe it will still be possible to have support for pyro models in parallel, as we have dataloading better handled now, and have a high level API, which doesn't exactly care if the model is pyro or pytorch or pytorch lightning.
